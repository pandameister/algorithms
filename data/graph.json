{
  "v0": [
    {
      "target": "v1",
      "weight": 37
    },
    {
      "target": "v0",
      "weight": 72
    },
    {
      "target": "v18",
      "weight": 92
    },
    {
      "target": "v6",
      "weight": 64
    },
    {
      "target": "v17",
      "weight": 49
    },
    {
      "target": "v8",
      "weight": 51
    },
    {
      "target": "v1",
      "weight": 95
    },
    {
      "target": "v14",
      "weight": 1
    }
  ],
  "v1": [
    {
      "target": "v0",
      "weight": 1
    },
    {
      "target": "v9",
      "weight": 83
    },
    {
      "target": "v14",
      "weight": 88
    },
    {
      "target": "v15",
      "weight": 31
    },
    {
      "target": "v6",
      "weight": 95
    },
    {
      "target": "v19",
      "weight": 67
    }
  ],
  "v2": [
    {
      "target": "v6",
      "weight": 2
    },
    {
      "target": "v19",
      "weight": 21
    },
    {
      "target": "v6",
      "weight": 14
    },
    {
      "target": "v19",
      "weight": 62
    },
    {
      "target": "v1",
      "weight": 13
    },
    {
      "target": "v12",
      "weight": 16
    }
  ],
  "v3": [
    {
      "target": "v7",
      "weight": 0
    }
  ],
  "v4": [
    {
      "target": "v18",
      "weight": 25
    },
    {
      "target": "v7",
      "weight": 22
    },
    {
      "target": "v7",
      "weight": 84
    },
    {
      "target": "v5",
      "weight": 69
    },
    {
      "target": "v1",
      "weight": 58
    },
    {
      "target": "v15",
      "weight": 28
    },
    {
      "target": "v2",
      "weight": 83
    },
    {
      "target": "v9",
      "weight": 38
    },
    {
      "target": "v16",
      "weight": 4
    },
    {
      "target": "v17",
      "weight": 32
    },
    {
      "target": "v14",
      "weight": 94
    }
  ],
  "v5": [
    {
      "target": "v2",
      "weight": 35
    },
    {
      "target": "v14",
      "weight": 74
    },
    {
      "target": "v18",
      "weight": 90
    },
    {
      "target": "v13",
      "weight": 2
    },
    {
      "target": "v7",
      "weight": 30
    },
    {
      "target": "v16",
      "weight": 13
    },
    {
      "target": "v6",
      "weight": 26
    },
    {
      "target": "v10",
      "weight": 64
    },
    {
      "target": "v12",
      "weight": 1
    },
    {
      "target": "v1",
      "weight": 15
    },
    {
      "target": "v5",
      "weight": 58
    },
    {
      "target": "v16",
      "weight": 44
    },
    {
      "target": "v19",
      "weight": 40
    },
    {
      "target": "v10",
      "weight": 87
    },
    {
      "target": "v0",
      "weight": 36
    },
    {
      "target": "v8",
      "weight": 28
    },
    {
      "target": "v11",
      "weight": 16
    }
  ],
  "v6": [
    {
      "target": "v9",
      "weight": 53
    },
    {
      "target": "v18",
      "weight": 78
    },
    {
      "target": "v19",
      "weight": 21
    },
    {
      "target": "v11",
      "weight": 3
    },
    {
      "target": "v11",
      "weight": 14
    }
  ],
  "v7": [
    {
      "target": "v3",
      "weight": 70
    }
  ],
  "v8": [
    {
      "target": "v8",
      "weight": 99
    },
    {
      "target": "v2",
      "weight": 4
    },
    {
      "target": "v10",
      "weight": 45
    },
    {
      "target": "v13",
      "weight": 89
    },
    {
      "target": "v5",
      "weight": 36
    },
    {
      "target": "v13",
      "weight": 53
    },
    {
      "target": "v17",
      "weight": 92
    },
    {
      "target": "v10",
      "weight": 52
    }
  ],
  "v9": [
    {
      "target": "v2",
      "weight": 44
    },
    {
      "target": "v6",
      "weight": 85
    },
    {
      "target": "v17",
      "weight": 26
    },
    {
      "target": "v15",
      "weight": 78
    },
    {
      "target": "v5",
      "weight": 53
    },
    {
      "target": "v8",
      "weight": 9
    },
    {
      "target": "v2",
      "weight": 53
    },
    {
      "target": "v4",
      "weight": 40
    },
    {
      "target": "v17",
      "weight": 76
    },
    {
      "target": "v2",
      "weight": 63
    },
    {
      "target": "v9",
      "weight": 44
    },
    {
      "target": "v13",
      "weight": 40
    }
  ],
  "v10": [
    {
      "target": "v0",
      "weight": 34
    },
    {
      "target": "v19",
      "weight": 99
    },
    {
      "target": "v13",
      "weight": 17
    },
    {
      "target": "v16",
      "weight": 68
    },
    {
      "target": "v3",
      "weight": 19
    },
    {
      "target": "v1",
      "weight": 23
    },
    {
      "target": "v4",
      "weight": 25
    },
    {
      "target": "v0",
      "weight": 20
    },
    {
      "target": "v2",
      "weight": 72
    },
    {
      "target": "v15",
      "weight": 16
    },
    {
      "target": "v18",
      "weight": 63
    },
    {
      "target": "v0",
      "weight": 24
    },
    {
      "target": "v6",
      "weight": 6
    },
    {
      "target": "v9",
      "weight": 20
    }
  ],
  "v11": [
    {
      "target": "v18",
      "weight": 1
    },
    {
      "target": "v13",
      "weight": 38
    },
    {
      "target": "v0",
      "weight": 31
    },
    {
      "target": "v17",
      "weight": 38
    },
    {
      "target": "v16",
      "weight": 45
    },
    {
      "target": "v17",
      "weight": 75
    },
    {
      "target": "v10",
      "weight": 0
    },
    {
      "target": "v5",
      "weight": 27
    },
    {
      "target": "v1",
      "weight": 79
    }
  ],
  "v12": [
    {
      "target": "v8",
      "weight": 80
    },
    {
      "target": "v15",
      "weight": 91
    }
  ],
  "v13": [
    {
      "target": "v9",
      "weight": 22
    },
    {
      "target": "v4",
      "weight": 22
    },
    {
      "target": "v8",
      "weight": 1
    },
    {
      "target": "v6",
      "weight": 66
    },
    {
      "target": "v12",
      "weight": 53
    },
    {
      "target": "v17",
      "weight": 54
    },
    {
      "target": "v1",
      "weight": 24
    },
    {
      "target": "v2",
      "weight": 20
    },
    {
      "target": "v14",
      "weight": 88
    },
    {
      "target": "v17",
      "weight": 47
    },
    {
      "target": "v12",
      "weight": 95
    },
    {
      "target": "v10",
      "weight": 84
    },
    {
      "target": "v10",
      "weight": 14
    },
    {
      "target": "v18",
      "weight": 87
    },
    {
      "target": "v10",
      "weight": 3
    },
    {
      "target": "v10",
      "weight": 34
    },
    {
      "target": "v12",
      "weight": 94
    },
    {
      "target": "v3",
      "weight": 27
    }
  ],
  "v14": [
    {
      "target": "v10",
      "weight": 57
    },
    {
      "target": "v7",
      "weight": 99
    },
    {
      "target": "v17",
      "weight": 25
    },
    {
      "target": "v0",
      "weight": 51
    },
    {
      "target": "v17",
      "weight": 14
    },
    {
      "target": "v4",
      "weight": 64
    },
    {
      "target": "v12",
      "weight": 75
    },
    {
      "target": "v16",
      "weight": 5
    },
    {
      "target": "v18",
      "weight": 14
    },
    {
      "target": "v4",
      "weight": 33
    },
    {
      "target": "v0",
      "weight": 95
    },
    {
      "target": "v6",
      "weight": 12
    },
    {
      "target": "v18",
      "weight": 28
    },
    {
      "target": "v7",
      "weight": 59
    },
    {
      "target": "v13",
      "weight": 79
    },
    {
      "target": "v10",
      "weight": 59
    }
  ],
  "v15": [
    {
      "target": "v0",
      "weight": 28
    },
    {
      "target": "v15",
      "weight": 42
    },
    {
      "target": "v11",
      "weight": 12
    },
    {
      "target": "v13",
      "weight": 97
    },
    {
      "target": "v4",
      "weight": 62
    },
    {
      "target": "v0",
      "weight": 62
    },
    {
      "target": "v4",
      "weight": 19
    },
    {
      "target": "v13",
      "weight": 95
    },
    {
      "target": "v19",
      "weight": 39
    },
    {
      "target": "v19",
      "weight": 79
    },
    {
      "target": "v5",
      "weight": 12
    },
    {
      "target": "v4",
      "weight": 84
    },
    {
      "target": "v14",
      "weight": 64
    },
    {
      "target": "v5",
      "weight": 9
    },
    {
      "target": "v7",
      "weight": 4
    },
    {
      "target": "v8",
      "weight": 85
    },
    {
      "target": "v13",
      "weight": 65
    },
    {
      "target": "v14",
      "weight": 42
    },
    {
      "target": "v14",
      "weight": 88
    }
  ],
  "v16": [
    {
      "target": "v17",
      "weight": 72
    },
    {
      "target": "v3",
      "weight": 44
    },
    {
      "target": "v12",
      "weight": 94
    },
    {
      "target": "v2",
      "weight": 41
    },
    {
      "target": "v9",
      "weight": 22
    }
  ],
  "v17": [
    {
      "target": "v7",
      "weight": 9
    },
    {
      "target": "v2",
      "weight": 56
    },
    {
      "target": "v11",
      "weight": 36
    },
    {
      "target": "v14",
      "weight": 43
    },
    {
      "target": "v10",
      "weight": 49
    },
    {
      "target": "v14",
      "weight": 92
    },
    {
      "target": "v15",
      "weight": 16
    },
    {
      "target": "v11",
      "weight": 39
    },
    {
      "target": "v11",
      "weight": 21
    },
    {
      "target": "v14",
      "weight": 38
    }
  ],
  "v18": [
    {
      "target": "v11",
      "weight": 87
    },
    {
      "target": "v16",
      "weight": 1
    },
    {
      "target": "v14",
      "weight": 76
    },
    {
      "target": "v4",
      "weight": 67
    },
    {
      "target": "v4",
      "weight": 40
    },
    {
      "target": "v14",
      "weight": 25
    },
    {
      "target": "v9",
      "weight": 80
    },
    {
      "target": "v19",
      "weight": 25
    },
    {
      "target": "v12",
      "weight": 93
    },
    {
      "target": "v12",
      "weight": 15
    },
    {
      "target": "v5",
      "weight": 66
    },
    {
      "target": "v1",
      "weight": 17
    },
    {
      "target": "v18",
      "weight": 79
    },
    {
      "target": "v15",
      "weight": 19
    },
    {
      "target": "v18",
      "weight": 69
    },
    {
      "target": "v12",
      "weight": 73
    },
    {
      "target": "v9",
      "weight": 93
    }
  ],
  "v19": [
    {
      "target": "v18",
      "weight": 62
    },
    {
      "target": "v14",
      "weight": 86
    },
    {
      "target": "v11",
      "weight": 70
    },
    {
      "target": "v17",
      "weight": 8
    },
    {
      "target": "v18",
      "weight": 2
    },
    {
      "target": "v12",
      "weight": 47
    },
    {
      "target": "v0",
      "weight": 34
    },
    {
      "target": "v6",
      "weight": 65
    },
    {
      "target": "v9",
      "weight": 32
    },
    {
      "target": "v14",
      "weight": 6
    },
    {
      "target": "v0",
      "weight": 69
    },
    {
      "target": "v15",
      "weight": 86
    },
    {
      "target": "v12",
      "weight": 40
    },
    {
      "target": "v7",
      "weight": 86
    }
  ]
}