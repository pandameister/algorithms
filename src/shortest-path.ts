interface ShortestPathInfo {
    [key: string]: {
        value: number;
        prevVertex: string | undefined;
    };
}
interface PathInfo {
    path: string[];
    value: number;
}

function initializeShortestPathInfo(graph: AdjacencyList, startVertex: string): ShortestPathInfo {
    const info: ShortestPathInfo = {};

    return Object.keys(graph).reduce((prev, curr) => {
        prev[curr] = {
            value: curr === startVertex ? 0 : Infinity,
            prevVertex: undefined
        };
        return prev;
    }, info);
}

function findVertextWithMinValue(vertices: string[], info: ShortestPathInfo): string {
    let minIdx: number | undefined;

    vertices.forEach((v, i) => {
        if (minIdx == null || info[v].value < info[vertices[minIdx]].value) {
            minIdx = i;
        }
    });

    return vertices.splice(minIdx!, 1)[0];
}

function computePath(info: ShortestPathInfo, targetVertex: string): string[] {
    const path = [];
    let curr: string | undefined = targetVertex;

    if (info[curr].value === Infinity) {
        return [];
    }

    while (curr != null) {
        path.push(curr);
        curr = info[curr].prevVertex;
    }

    return path.reverse();
}

function computeShortestPathInfo(graph: AdjacencyList, startVertex: string, endVertex?: string): ShortestPathInfo {
    const info = initializeShortestPathInfo(graph, startVertex);
    const unvisited = Object.keys(graph);

    while (unvisited.length > 0) {
        const vertex = findVertextWithMinValue(unvisited, info);

        // we can stop if reached the endVertex
        if (vertex === endVertex) {
            break;
        }

        const edges = graph[vertex];
        const vertexInfo = info[vertex];

        for (const edge of edges) {
            const targetInfo = info[edge.target];

            const targetValue = vertexInfo.value + edge.weight;
            if (targetInfo.value > targetValue) {
                targetInfo.value = targetValue;
                targetInfo.prevVertex = vertex;
            }
        }
    }

    return info;
}

/**
 * Computes the shortest path(s) for a graph from a given vertex
 *
 * @param graph - The graph
 * @param startVertex - The start vertex
 * @param endVertex - The end vertex. If no end vertex is provided, all paths are computed
 * @returns The shortest path or an array of shortest paths
 */
export function shortestPath(graph: AdjacencyList, startVertex: string, endVertex?: string): PathInfo | PathInfo[] {
    if (endVertex != null) {
        const info = computeShortestPathInfo(graph, startVertex, endVertex);
        return { path: computePath(info, endVertex), value: info[endVertex].value };
    } else {
        const info = computeShortestPathInfo(graph, startVertex);

        return Object.keys(graph)
            .map((endVertex) => {
                return { path: computePath(info, endVertex), value: info[endVertex].value };
            })
            .filter((path) => path.value !== Infinity);
    }
}
