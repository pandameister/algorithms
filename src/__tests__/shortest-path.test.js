import graph from '../../data/graph.simple.json';
import complexGraph from '../../data/graph.json';
import { shortestPath } from '../shortest-path';

function assertShortestPath(src, dst, expected, testReverse = true) {
    let path = shortestPath(graph, src, dst);
    expect(path).toEqual(expected);

    if (testReverse) {
        path = shortestPath(graph, dst, src);
        expected = {...expected, path: [...expected.path].reverse()}
        expect(path).toEqual(expected);
    }
}

describe('Test Simple Graph', () => {
    test('shortest path', () => {
        assertShortestPath('A', 'C', {
            path: ['A', 'D', 'E', 'C'],
            value: 7
        });

        assertShortestPath('A', 'A', {
            path: ['A'],
            value: 0
        });

        assertShortestPath('A', 'B', {
            path: ['A', 'D', 'B'],
            value: 3
        });

        assertShortestPath('A', 'F', {
            path: [],
            value: Infinity
        }, false);
    });

    test('all shortest path', () => {
        const paths = shortestPath(graph, 'A');
        expect(paths).toMatchSnapshot();
    });
});

describe('Test Complex Graph', () => {
    test('all shortest path', () => {
        const paths = shortestPath(complexGraph, 'v0');
        expect(paths).toMatchSnapshot();
    });
});