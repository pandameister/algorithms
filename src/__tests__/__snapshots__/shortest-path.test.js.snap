// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Complex Graph all shortest path 1`] = `
Array [
  Object {
    "path": Array [
      "v0",
    ],
    "value": 0,
  },
  Object {
    "path": Array [
      "v0",
      "v14",
      "v18",
      "v1",
    ],
    "value": 32,
  },
  Object {
    "path": Array [
      "v0",
      "v14",
      "v6",
      "v11",
      "v10",
      "v13",
      "v8",
      "v2",
    ],
    "value": 38,
  },
  Object {
    "path": Array [
      "v0",
      "v14",
      "v6",
      "v11",
      "v10",
      "v3",
    ],
    "value": 35,
  },
  Object {
    "path": Array [
      "v0",
      "v14",
      "v4",
    ],
    "value": 34,
  },
  Object {
    "path": Array [
      "v0",
      "v14",
      "v17",
      "v15",
      "v5",
    ],
    "value": 40,
  },
  Object {
    "path": Array [
      "v0",
      "v14",
      "v6",
    ],
    "value": 13,
  },
  Object {
    "path": Array [
      "v0",
      "v14",
      "v17",
      "v7",
    ],
    "value": 24,
  },
  Object {
    "path": Array [
      "v0",
      "v14",
      "v6",
      "v11",
      "v10",
      "v13",
      "v8",
    ],
    "value": 34,
  },
  Object {
    "path": Array [
      "v0",
      "v14",
      "v16",
      "v9",
    ],
    "value": 28,
  },
  Object {
    "path": Array [
      "v0",
      "v14",
      "v6",
      "v11",
      "v10",
    ],
    "value": 16,
  },
  Object {
    "path": Array [
      "v0",
      "v14",
      "v6",
      "v11",
    ],
    "value": 16,
  },
  Object {
    "path": Array [
      "v0",
      "v14",
      "v18",
      "v12",
    ],
    "value": 30,
  },
  Object {
    "path": Array [
      "v0",
      "v14",
      "v6",
      "v11",
      "v10",
      "v13",
    ],
    "value": 33,
  },
  Object {
    "path": Array [
      "v0",
      "v14",
    ],
    "value": 1,
  },
  Object {
    "path": Array [
      "v0",
      "v14",
      "v17",
      "v15",
    ],
    "value": 31,
  },
  Object {
    "path": Array [
      "v0",
      "v14",
      "v16",
    ],
    "value": 6,
  },
  Object {
    "path": Array [
      "v0",
      "v14",
      "v17",
    ],
    "value": 15,
  },
  Object {
    "path": Array [
      "v0",
      "v14",
      "v18",
    ],
    "value": 15,
  },
  Object {
    "path": Array [
      "v0",
      "v14",
      "v6",
      "v19",
    ],
    "value": 34,
  },
]
`;

exports[`Test Simple Graph all shortest path 1`] = `
Array [
  Object {
    "path": Array [
      "A",
    ],
    "value": 0,
  },
  Object {
    "path": Array [
      "A",
      "D",
      "B",
    ],
    "value": 3,
  },
  Object {
    "path": Array [
      "A",
      "D",
      "E",
      "C",
    ],
    "value": 7,
  },
  Object {
    "path": Array [
      "A",
      "D",
    ],
    "value": 1,
  },
  Object {
    "path": Array [
      "A",
      "D",
      "E",
    ],
    "value": 2,
  },
]
`;
